File: src/claw_mem/reflection/orchestrator.py
"""
claw-mem v2.9.1 - Reflection Orchestrator

Coordinates the full reflection pipeline:
  1. Collect recent observations from memory
  2. Synthesize into beliefs
  3. Track belief changes
  4. Store results back to memory

Reflection is triggered:
  - On command: memory.reflect()
  - On schedule: automatically when enough new observations accumulate
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, List, Optional

from .synthesizer import BeliefSynthesizer, Belief, Observation, SynthesizerConfig
from .belief_tracker import BeliefTracker, BeliefVersion


@dataclass
class ReflectionResult:
    """Result of a reflection operation."""
    observations: List[Observation] = field(default_factory=list)
    beliefs: List[Belief] = field(default_factory=list)
    new_beliefs: List[Belief] = field(default_factory=list)
    updated_beliefs: List[Belief] = field(default_factory=list)
    timestamp: str = ""
    summary: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "observations_count": len(self.observations),
            "beliefs_total": len(self.beliefs),
            "beliefs_new": len(self.new_beliefs),
            "beliefs_updated": len(self.updated_beliefs),
            "timestamp": self.timestamp,
            "summary": self.summary,
            "observations": [o.to_dict() for o in self.observations[:10]],
            "beliefs": [b.to_dict() for b in self.beliefs[:10]],
        }


class ReflectionOrchestrator:
    """Orchestrate the reflection observation→belief pipeline.

    Usage:
        orch = ReflectionOrchestrator()
        result = orch.reflect(memories, user_id="user123")
        print(f"Synthesized {len(result.beliefs)} beliefs")
    """

    def __init__(self, config: Optional[SynthesizerConfig] = None):
        self.synthesizer = BeliefSynthesizer(config)
        self.tracker = BeliefTracker()
        self._last_reflection_at: Optional[str] = None
        self._reflection_count = 0

    def reflect(
        self, memories: List[Dict[str, Any]], user_id: str = "",
        force: bool = False,
    ) -> ReflectionResult:
        """Execute a full reflection cycle.

        Args:
            memories: Recent memory records to reflect on
            user_id: User identifier
            force: Force reflection even if not enough data

        Returns:
            ReflectionResult with observations and beliefs
        """
        now = datetime.utcnow().isoformat()
        self._reflection_count += 1
        self._last_reflection_at = now

        # Step 1: Extract observations
        observations = self.synthesizer.extract_observations(memories)

        # Step 2: Synthesize beliefs
        beliefs = self.synthesizer.synthesize(observations, user_id)

        # Step 3: Track changes (new vs updated)
        new_beliefs = []
        updated_beliefs = []

        for belief in beliefs:
            existing = self.tracker.get_current(belief.id)
            if existing:
                # Check if actually changed
                if existing.statement != belief.statement:
                    self.tracker.update(belief.id, belief.statement, belief.confidence)
                    updated_beliefs.append(belief)
            else:
                self.tracker.record(belief.id, belief.statement, belief.confidence)
                new_beliefs.append(belief)

        # Step 4: Build summary
        summary = (
            f"Reflection #{self._reflection_count}: "
            f"{len(observations)} observations → "
            f"{len(beliefs)} beliefs "
            f"({len(new_beliefs)} new, {len(updated_beliefs)} updated)"
        )

        return ReflectionResult(
            observations=observations,
            beliefs=beliefs,
            new_beliefs=new_beliefs,
            updated_beliefs=updated_beliefs,
            timestamp=now,
            summary=summary,
        )

    def get_beliefs(self, include_history: bool = False) -> List[Dict[str, Any]]:
        """Get all current beliefs.

        Args:
            include_history: Include version history

        Returns:
            List of belief dicts
        """
        beliefs = []
        for belief_id in self.tracker.get_all_ids():
            version = self.tracker.get_current(belief_id)
            if version:
                b = version.to_dict()
                if include_history:
                    b["history"] = [v.to_dict() for v in self.tracker.get_history(belief_id)]
                beliefs.append(b)
        return beliefs

    def get_reflection_stats(self) -> Dict[str, Any]:
        """Get reflection statistics."""
        return {
            "reflection_count": self._reflection_count,
            "last_reflection_at": self._last_reflection_at,
            "total_beliefs": len(self.tracker._store),
            "total_versions": sum(
                len(v) for v in self.tracker._store.values()
            ),
        }

File: src/reflection/orchestrator.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Reflection Orchestrator
 *
 * Coordinates the full reflection pipeline:
 *   1. Collect recent observations from memory
 *   2. Synthesize into beliefs
 *   3. Track belief changes
 *   4. Store results back to memory
 */

import { BeliefTracker } from "./belief_tracker";
import {
  BeliefSynthesizer,
  Belief,
  Observation,
  SynthesizerConfig,
} from "./synthesizer";

// ── ReflectionResult ──────────────────────────────────────────────────

export interface ReflectionResult {
  observations: Observation[];
  beliefs: Belief[];
  new_beliefs: Belief[];
  updated_beliefs: Belief[];
  timestamp: string;
  summary: string;
}

// ── ReflectionOrchestrator ────────────────────────────────────────────

export class ReflectionOrchestrator {
  synthesizer: BeliefSynthesizer;
  tracker: BeliefTracker;
  private _lastReflectionAt: string | null = null;
  private _reflectionCount = 0;

  constructor(config?: Partial<SynthesizerConfig>) {
    this.synthesizer = new BeliefSynthesizer(config);
    this.tracker = new BeliefTracker();
  }

  /**
   * Execute a full reflection cycle.
   *
   * @param memories - Recent memory records to reflect on
   * @param userId - User identifier
   * @param force - Force reflection even if not enough data
   * @returns ReflectionResult with observations and beliefs
   */
  reflect(
    memories: Record<string, unknown>[],
    userId: string = "",
    force: boolean = false,
  ): ReflectionResult {
    const now = new Date().toISOString();
    this._reflectionCount++;
    this._lastReflectionAt = now;

    // Step 1: Extract observations
    const observations = this.synthesizer.extract_observations(memories);

    // Step 2: Synthesize beliefs
    const beliefs = this.synthesizer.synthesize(observations, userId);

    // Step 3: Track changes (new vs updated)
    const newBeliefs: Belief[] = [];
    const updatedBeliefs: Belief[] = [];

    for (const belief of beliefs) {
      const existing = this.tracker.get_current(belief.id);
      if (existing) {
        if (existing.statement !== belief.statement) {
          this.tracker.update(belief.id, belief.statement, belief.confidence);
          updatedBeliefs.push(belief);
        }
      } else {
        this.tracker.record(belief.id, belief.statement, belief.confidence);
        newBeliefs.push(belief);
      }
    }

    // Step 4: Build summary
    const summary =
      `Reflection #${this._reflectionCount}: ` +
      `${observations.length} observations \u2192 ` +
      `${beliefs.length} beliefs ` +
      `(${newBeliefs.length} new, ${updatedBeliefs.length} updated)`;

    return {
      observations,
      beliefs,
      new_beliefs: newBeliefs,
      updated_beliefs: updatedBeliefs,
      timestamp: now,
      summary,
    };
  }

  /**
   * Get all current beliefs.
   *
   * @param includeHistory - Include version history
   * @returns List of belief dicts
   */
  get_beliefs(includeHistory: boolean = false): Record<string, unknown>[] {
    const beliefs: Record<string, unknown>[] = [];
    for (const beliefId of this.tracker.get_all_ids()) {
      const version = this.tracker.get_current(beliefId);
      if (version) {
        const b: Record<string, unknown> = {
          belief_id: version.belief_id,
          statement: version.statement,
          confidence: version.confidence,
          version: version.version,
          created_at: version.created_at,
        };
        if (includeHistory) {
          b.history = this.tracker
            .get_history(beliefId)
            .map((v) => ({
              belief_id: v.belief_id,
              statement: v.statement,
              confidence: v.confidence,
              version: v.version,
              created_at: v.created_at,
              previous_statement: v.previous_statement,
            }));
        }
        beliefs.push(b);
      }
    }
    return beliefs;
  }

  /**
   * Get reflection statistics.
   *
   * @returns Dict with reflection stats
   */
  get_reflection_stats(): Record<string, unknown> {
    return {
      reflection_count: this._reflectionCount,
      last_reflection_at: this._lastReflectionAt,
      total_beliefs: this.tracker.count_beliefs(),
      total_versions: this.tracker.count_versions(),
    };
  }
}
